sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,530 @@
+  task_sequence(task_sequence &&) = delete;
+  task_sequence &operator=(task_sequence &&) = delete;
+
+
+  void async(__unspecified__... args);
+
+  __unspecified__ get();
+
+  template<typename propertyT>
+  static constexpr bool has_property();
+
+  // The return type is an unspecified internal class used to represent
+  // instances of propertyT
+  template<typename propertyT>
+  static constexpr /*unspecified*/ get_property();
+
+  ~task_sequence();
+}
+
+} // namespace sycl::ext::intel::experimental
+
+----
+
+`task_sequence` is a class template, parameterized by an `auto` reference to a
+Callable `f`, and a list of properties _PropertyListT_. The Callable `f` defines
+the asynchronous task to be associated with the `task_sequence`, and requiring
+an auto reference ensures that each `f` is statically resolvable at compile
+time. Static resolvability by the compiler is desirable when compiling for
+spatial architectures as it can enable the generation of more efficient
+hardware. Templating a task_sequence on a non-Callable is malformed.
+_PropertyListT_ enables properties to be associated with a `task_sequence`
+
+[cols="1,1"]
+|===
+|Member functions of the `task_sequence` class|Description
+
+|void async(\\__unspecified__... args)
+|Available only in device functions. +
+ +
+ Asynchronously invoke the task_sequence's associated Callable `f` with `args`
+ Calling `async` on a task_sequence with mismatched argument types with respect
+ to the Callable `f`'s argument types is malformed.
+
+|\\__unspecified__ get()
+|Available only in device functions. +
+ +
+ Retrieves a return value from the task_sequence, which is the same return type
+ as the associated Callable `f`'s return type.
+|===
+
+By calling `async` on a `task_sequence` object more than once, the user implies
+that those invocations of `f` can be run in parallel. The implementation is,
+however, not obligated to run the invocations in parallel except in so far as is
+necessary to meet the forward progress guarantees outlined in the section on
+Progress Guarantees.
+
+The `async` function call is non-blocking in that it may return before the
+asynchronous `f` invocation completes executing, and potentially before `f` even
+begins executing (return from the `async` provides no implicit information on
+the execution status of `f`).
+
+An asynchronous invocation of `f`, upon the end of its execution of `f`, will
+produce a result. At that point, that particular asynchronous invocation is now
+considered to be complete, and the result of that invocation is now ready to be
+retrieved by a `get` call. The result of each completed invocation is stored in
+a results data structure until the appropriate `get` call retrieves it. This
+results data structure will keep the results in the same order in which their
+corresponding `async` invocations were invoked. The `get` call retrieves results
+from this results data structure. The `get` call blocks if there is no result to
+retrieve.
```

Comment:
I worry that with the current wording someone might still think that results can be returned out of order so long as the earlier asyncs haven't completed yet, i.e. that only those asyncs that are done are ordered.  I think we could clarify by saying:

```suggestion
from this results data structure. The `get` call blocks if the result for the oldest 'async' is not in the results data structure.  
```

Reply:
Now that I think about it more, I think this might be indicative of a bigger problem.  With the way we currently define the behaviour of the results data structure and our forward progress guarantees, I worry that a user has to size their results data structure equal to their maximum number of outstanding invocations (i.e. their invocation_capacity) if they want to guarantee no hangs which is not our intent.  Consider this case:

-invocation_capacity of 2
-response capacity of 1
Code:
```
task_sequence<...> t;
t.async();
t.async();
t.get();

The get() must return the result of the first async.  Let's say the 2nd thread get strengthened to have a concurrent forward progress guarantee.  When it completes, it stores its result in the results data structure, taking up the one slot available based on a response capacity of 1 and delegates its forward progress guarantee to the first thread.  The first thread completes but can't store its result into the results data structure because it's full.  As a result the get() never returns because it needs the result from that first thread.  

I think to get out of this mess we may need to tweak the definition of response_capacity slightly to indicate that a response_capacity of N means you must have sufficient capacity that the _oldest outstanding N invocations_ can fit in the results data structure.  

Such a definition would require implementations to either:
1. Keep invocations in order if they only want to use a capacity of N
2. If they want to allow invocations to get out of order they'd have to add extra capacity to account for the maximum amount of out-of-orderness they allowed.

Reply:
I've changed the wording of response capacity to reflect your first comment for now. Awaiting Abhishek's opinions.

Reply:
I like the idea of removing the ordering guarantee and simplifying the spec

Reply:
I think we are going to stick with guaranteed return order for now to ensure emulator and simulator behaviour match, with the option to add a property later on to allow for out of order returns.